pkg/writer/serializer_cdx.go

```diff
@@ -48,25 +94,28 @@ func (s *SerializerCDX) Serialize(opts options.Options, bom *sbom.Document) (int
 		}
 
 		if comp.BOMRef == "" {
-			refless = append(refless, comp)
+			s.reflessList = append(s.reflessList, comp)
 		} else {
-			components[comp.BOMRef] = comp
+			s.componentsDict[comp.BOMRef] = comp
 		}
 	}
+	return nil
+}
 
-	rootDict := map[string]struct{}{}
-	addedDict := map[string]struct{}{}
+func (s *SerializerCDX) root(bom *sbom.Document) (*cdx.Component, error) {
+	var rootComp *cdx.Component
+	found := false
 
 	// First, assign the top level nodes
 	if bom.RootElements != nil && len(bom.RootElements) > 0 {
 		for _, id := range bom.RootElements {
-			rootDict[id] = struct{}{}
+			s.rootDict[id] = struct{}{}
 			// Search for the node and add it
 			for _, n := range bom.Nodes {
 				if n.Id == id {
-					rootComp := s.nodeToComponent(n)
-					doc.Metadata.Component = rootComp
-					addedDict[id] = struct{}{}
+					rootComp = s.nodeToComponent(n)
+					found = true
+					s.addedDict[id] = struct{}{}
 				}
 			}
```

Comment:
Most of this logic can be removed in favor of the API functions I've started to build around the autogenerated objects, starting with https://github.com/bom-squad/protobom/pull/20 (see `document.GetRootNodes()` for a start)

Reply:
OK i added a comment to refactor the func when the PR is merged.
I also just noticing the rootDict is not used by anyone in the code so it can be simply removed for now anway